apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClientBuilder.cs

```diff
@@ -0,0 +1,153 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at 
+//
+// https://www.apache.org/licenses/LICENSE-2.0 
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.PubSub.V1;
+
+/// <summary>
+/// Builder class for <see cref="PublisherClient"/> to provide simple configuration of credentials,
+/// endpoint, client count, publication settings etc.
+/// </summary>
+public sealed class PublisherClientBuilder : ClientBuilderBase<PublisherClient>
+{
+    private static readonly GrpcChannelOptions s_unlimitedSendReceiveChannelOptions = GrpcChannelOptions.Empty
+        .WithMaxReceiveMessageSize(int.MaxValue)
+        .WithMaxSendMessageSize(int.MaxValue)
+        // Set max metadata size to 4 MB i.e., 4194304 bytes.
+        .WithCustomOption("grpc.max_metadata_size", 4194304);
+
+    /// <summary>
+    /// The name of the topic that the publisher publishes to.
+    /// This must be non-null by the time <see cref="Build"/> or <see cref="BuildAsync(CancellationToken)"/> is called.
+    /// </summary>
+    public TopicName TopicName
+    {
+        get; set;
+    }
+
+    private int? _clientCount;
+
+    /// <summary>
+    /// The number of <see cref="PublisherServiceApiClient"/>s to create and use within a <see cref="PublisherClient"/> instance.
+    /// If this is null, the number of clients created will depend on the processor count.
+    /// </summary>
+    public int? ClientCount
+    {
+        get => _clientCount;
+        set
+        {
+            if (value is int count)
+            {
+                // Fairly arbitrary upper limit.
+                _clientCount = GaxPreconditions.CheckArgumentRange(count, nameof(value), 1, 256);
```

Comment:
Maybe check this in Validate? Sorry for not saying anything before.

And same in SubscriberClientBuilder.

Reply:
Previously I had it in validate, I think :)

The expectation is that it's more useful in the property setter, as it's immediately wrong, and it's never useful to explicitly set it to an invalid value. (Compare that with TopicName, which is null to start with, and has to be non-null in the end, but if we made `builder.TopicName = builder.TopicName;` throw an exception, that would be odd.)

This is similar to what we do in Storage, e.g. https://github.com/googleapis/google-cloud-dotnet/blob/main/apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UploadObjectOptions.cs